### PR TITLE
API for neo-OBS

### PR DESCRIPTION
**Pull request by HuseinA**
- [x] endpoint

- [x] response status code

- [x] exception

- [x] restructure directory

- [x] config

- [x] docker

### CODE REVIEW

File: obs/api/app/controllers/api/config.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
+from obs.libs import config
+from app.helpers.rest import response
+from flask import request, jsonify
+from flask_restful import Resource, reqparse
+
+def open_config():
+    data={}
```
Comment (riszkymf) [inline obs/api/app/controllers/api/config.py]: jangan lupa ditutup lagi file nya.
https://www.pythonforbeginners.com/cheatsheet/python-file-handling
coba pake with
https://stackoverflow.com/questions/9282967/how-to-open-a-file-using-the-open-with-statement

File: obs/api/app/controllers/api/bucket.py [hunk 2]
```diff
@@ -22,143 +23,213 @@ def get(self):
...
+            return response(500)
 
     def post(self, bucket_name):
-        bucket.create_bucket(auth=get_plain_auth(), bucket_name=bucket_name)
-        return response(201)
+        try:
+            bucket.create_bucket(auth=get_plain_auth(), bucket_name=bucket_name)
+            return response(201)
+        except Exception:
+            return response(500)
```
Comment (riszkymf) [inline obs/api/app/controllers/api/bucket.py]: kalo bisa dikasih informasi di response error nya kenapa.

contoh: 

except Exception as e:
    return response(500,message=str(e))

File: CHANGELOG.rst [hunk 3]
```diff
@@ -1,6 +1,14 @@
 Unreleased
 ==========
 
+0.1.1 (2020-01-13)
```
Comment (azzamsa) [inline CHANGELOG.rst]: is it part of obs web API?

File: obs/api/app/controllers/api/admin.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+                users = user.info(get_client(), args["userId"], args["groupId"])
+                if "reason" in users:
+                    return response(users["status_code"], message=users["reason"])
+                return response(200, data=users)
+
+            user_type = args["user_type"] if args["user_type"] else "all"
+            user_status = args["user_status"] if args["user_status"] else "active"
+            limit = args["limit"] if args["limit"] else ""
+
+            list = user.list_user(
```
Comment (azzamsa) [inline obs/api/app/controllers/api/admin.py]: Can you use meaningful variable name? and not using reserved variable?

File: obs/api/app/controllers/api/bucket.py [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+def get_resources(access_key, secret_key):
+    config.load_config_file()
+    endpoint = auth.get_endpoint()
+    sess = boto3.Session(aws_access_key_id=access_key, aws_secret_access_key=secret_key)
+    s3_resource = sess.resource("s3", endpoint_url=endpoint)
+    return s3_resource
+
+
+def get_plain_auth(access_key, secret_key):
+    config.load_config_file()
```
Comment (azzamsa) [inline obs/api/app/controllers/api/bucket.py]: still use config file?

File: obs/api/app/controllers/api/admin.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+        parser.add_argument("suspend", type=str, required=True)
+        args = parser.parse_args()
+
+        try:
+            msg = "suspended" if args["suspend"] == "true" else "unsuspended"
+            users = user.info(get_client(), args["userId"], args["groupId"])
+            if users["active"] == f"{not strtobool(args['suspend'])}".lower():
+                return response(400, f"User already {msg}")
+
+            del users["canonicalUserId"]
```
Comment (azzamsa) [inline obs/api/app/controllers/api/admin.py]: Would you like to explain why, you need to delete that element?

File: obs/api/app/controllers/api/config.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
+from obs.libs import config
+from app.helpers.rest import response
+from flask import request, jsonify
+from flask_restful import Resource, reqparse
+
+
```
Comment (azzamsa) [inline obs/api/app/controllers/api/config.py]: do you really need those operation? changing the config file via WEB API?

File: obs/api/requirements.txt [hunk 8]
```diff
@@ -0,0 +1,19 @@
+Click==7.0
```
Comment (azzamsa) [inline obs/api/requirements.txt]: do you really need click for web API?

File: obs/api/requirements.txt [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+tabulate==0.8.3
+xmltodict==0.12.0
+requests==2.22.0
+requests-aws4auth==0.9
+PyYAML==5.1.2
+Flask_Script==2.0.6
+Flask==1.0.2
+Flask_Cors==3.0.7
+Flask_RESTful==0.3.7
+coloredlogs==10.0
```
Comment (azzamsa) [inline obs/api/requirements.txt]: are you really using xmldtodict and coloredlogs?

**Comment by azzamsa** [comment]
I see some unused things. Please use flake8 in your editor (as plugin) or run it manually.

File: obs/api/app/controllers/api/config.py (hunk 1 above)
Comment (HuseinA) [reply obs/api/app/controllers/api/config.py]: file deleted cause we did't use config in local anymore

File: obs/api/app/controllers/api/bucket.py (hunk 5 above)
Comment (HuseinA) [reply obs/api/app/controllers/api/bucket.py]: deleted

File: obs/api/docker-compose.yml [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+    ports:
+        - "5000:5000"
+    environment:
+      - FLASK_APP=autoapp.py
+      - FLASK_ENV=production
+      - GUNICORN_WORKERS=4
+      - FLASK_DEBUG=0
+      - APP_HOST=0.0.0.0
+      - APP_PORT=5000
+      - OBS_USER_URL=s3-stage.biznetgio.net
```
Comment (azzamsa) [inline obs/api/docker-compose.yml]: just leave internal info/url such this empty.

File: obs/api/app/libs/utils.py [hunk 11]
```diff
@@ -0,0 +1,52 @@
+import requests
+import json
+import os
```
Comment (azzamsa) [inline obs/api/app/libs/utils.py]: have you check it using flake8? it's unused import

File: obs/api/app/libs/utils.py [hunk 12]
```diff
@@ -0,0 +1,52 @@
+import requests
```
Comment (azzamsa) [inline obs/api/app/libs/utils.py]: utils.py shouldn't be under utils/ dir? (not crucial, you can ignore this.)

File: obs/api/app/controllers/api/admin.py [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+        args = parser.parse_args()
+
+        try:
+            msg = "suspended" if args["suspend"] == "true" else "unsuspended"
+            users = user.info(get_client(), args["userId"], args["groupId"])
+            if users["active"] == f"{not strtobool(args['suspend'])}".lower():
+                return response(400, f"User already {msg}")
+
+            # canocicalUserId can't be included when updating user
+            del users["canonicalUserId"]
```
Comment (azzamsa) [inline obs/api/app/controllers/api/admin.py]: not pythonic way. but as you wish.

File: obs/api/app/controllers/api/admin.py [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+
+class cred_api(Resource):
+    def get(self):
+        parser = reqparse.RequestParser()
+        parser.add_argument("userId", type=str, required=True)
+        parser.add_argument("groupId", type=str, required=True)
+        args = parser.parse_args()
+
+        try:
+            list = credential.list(get_client(), args["userId"], args["groupId"])
```
Comment (azzamsa) [inline obs/api/app/controllers/api/admin.py]: avoid using reserved word, it's hard to debug in the end of the day

File: obs/api/app/controllers/api/bucket.py [hunk 15]
```diff
@@ -0,0 +1,402 @@
...
+    return s3_resource
+
+
+def get_plain_auth(access_key, secret_key):
+    auth = AWS4Auth(access_key, secret_key, "eu-west-1", "s3")
+    return auth
+
+
+class list(Resource):
+    def get(self, prefix=""):
```
Comment (azzamsa) [inline obs/api/app/controllers/api/bucket.py]: this function is too long. you can separate/make it more shorter.

File: obs/api/app/controllers/api/bucket.py [hunk 16]
```diff
@@ -0,0 +1,402 @@
...
+                    get_resources(args["access_key"], secret_key), bucket_name, prefix
+                )
+                objects = []
+                if buckets["CommonPrefixes"]:
+                    for prefix in buckets["CommonPrefixes"]:
+                        objects.append({"directory": f"{prefix['Prefix']}"})
+
+                if buckets["Contents"]:
+                    for content in buckets["Contents"]:
+                        last_modified = content["LastModified"]
```
Comment (azzamsa) [inline obs/api/app/controllers/api/bucket.py]: unused ?

**Pull request merged**